### ISSUE TITLE
how do i check the file before upload?

### ISSUE THREAD

**Issue by yuanyuanweiwu** [flowjs/flow.js#271]
Now I have a requirement to call the background interface to detect before uploading the file. If the file already exists, cancel the upload and prompt, no upload file. Which method or configuration property do I use to perform this judgment? Thank you

**Comment by MulderNick** [comment]
you an use the events "fileAdded" and "filesAdded"

const flow = new Flow()
flow.on("filesAdded", function(flowFiles) {
  // ToDo implement duplication check
});

you can check the files there, remove all duplicated files from the flow instance via flowFile.cancel() and then after that start the upload with flow.resume() or flow.upload()

**Comment by madsnow** [comment]
Is it possible to perform an async operation in fileAdded?

**Comment by AidasK** [comment]
No, it's not. You have to reject a file and add it later instead

**Comment by drzraf** [comment]
Highly related to the discussion initiated at https://github.com/flowjs/flow.js/issues/319
- filesAdded as a **hook** to interact with the process (not async) like a filter
vs
- filesAdded as an **event/notification** providing information to a (possibly async) listener acting on a side-channel

The last comment from madsnow was look for an `async filter` withing the general processing of `filesAdded`. I can't figure how this could be done correctly and don't think this should be supported (at least for now).

**Comment by drzraf** [comment]
I plan to add a `preFilterFiles` hook (`filesAdded` staying as a simple notification event), in a future PR.
This hook would expect a Promise to be returned and block further processing (the call to `addFiles`, (`async` too)) until the hook's promise returns.

I don't know if it's the best JS-development pattern, but it'd resolve that problem (and the one caused by `filesAdded`) using a reproducible pattern and avoid adding `state variable` into Flow or FlowFile.

**Comment by AidasK** [comment]
@drzraf sounds awesome 👍

**Comment by drzraf** [comment]
@madsnow  : You could have a look at https://github.com/flowjs/flow.js/pull/329 and see if
```js
await asyncAddFile(yourfile, null, async (flowFile, event) => {
 // whatever you want before initialization
});
```
would fit (where the 3rd parameter is an async equivalent of `initFileFn`).

Note:
- `flowFile` isn't yet fully bootstrapped at that moment (but FlowFile `_bootstrap()` isn't rocket-science either)
- `event` may indicates whether its a _retried_ attempt.
